lightning/impl/sag_fast.pyx

```diff
@@ -19,6 +20,40 @@ from lightning.impl.dataset_fast cimport RowDataset
 from lightning.impl.sgd_fast cimport LossFunction
 
 
+cdef class ProxFunction:
+
+    cdef void apply_prox(self, double* x, int x_size):
+        raise NotImplementedError()
+
+
+cdef class L1Prox(ProxFunction):
+
+    cdef double l1
+
+    def __init__(self, double l1=0.):
+        self.l1 = l1
+
+    cdef void apply_prox(self, double* x, int x_size):
+        cdef int j
+        for j in xrange(x_size):
+            x[j] = np.sign(x[j]) * np.abs(x[j] - self.l1) * (np.abs(x[j]) > self.l1)
+        return
+
+
+cdef class L2Prox(ProxFunction):
```

Comment:
There is no need to add L2Prox since SAGA already has a parameter for the L2 regularization (it is included in the loss since its smooth).